File: 2023-12-21/leetcode_297.py
from collections import deque

class Codec:

    def serialize(self, root: TreeNode) -> str:
        data = []
        
        def dfs(node):
            if not node:
                data.append('null')
                return
        
            data.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
        
        dfs(root)
        return ' '.join(data)

    def deserialize(self, data : str) -> TreeNode:
        dataQ = deque(data.split())

        def dfs(q):
            if not len(q):
                return
            
            val = q.popleft()
            if val == 'null':
                return None
            
            node = TreeNode(val)
            node.left = dfs(q)
            node.right = dfs(q)
            return node
        
        return dfs(dataQ)

File: leetcode/leetcode_297.js
const serialize = (root) => {
  const data = [];

  const dfs = (node) => {
    if (node === null) {
      data.push('null');
      return;
    }

    data.push(node.val);
    dfs(node.left);
    dfs(node.right);
  };

  dfs(root);
  return data.join(' ');
};

const deserialize = (data) => {
  dataList = data.split(' ');

  const dfs = (list) => {
    if (list.length === 0) {
      return;
    }

    const val = list.shift();
    if (val === 'null') {
      return null;
    }

    const node = new TreeNode(val);
    node.left = dfs(list);
    node.right = dfs(list);
    return node;
  };

  return dfs(dataList);
};
